pkg/volume/empty_dir/quota.go

```diff
@@ -0,0 +1,128 @@
+package empty_dir
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"os/exec"
+	"strings"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+)
+
+// An object for applying quota to an empty dir volume.
+type quotaApplicator interface {
+	// Apply the quota to the given EmptyDir path:
+	Apply(dir string, medium api.StorageMedium, pod *api.Pod) error
+}
+
+type xfsQuotaApplicator struct {
+}
+
+func (xqa *xfsQuotaApplicator) Apply(dir string, medium api.StorageMedium, pod *api.Pod, fsGroup *int64, quota int) error {
+	if medium == api.StorageMediumMemory {
+		glog.V(3).Infof("skipping quota application due to memory storage medium.")
+		return nil
+	}
+	if !xqa.isXfs(dir) {
+		glog.V(3).Infof("%s is not on an XFS filesystem, skipping quota application.")
+		return nil
+	}
+	if fsGroup == nil {
+		// This indicates the operation matched an SCC with FSGroup strategy RunAsAny.
+		// Not an error condition.
+		glog.V(3).Infof("Unable to apply XFS quota, no FSGroup specified.")
+		return nil
+	}
+
+	volDevice, err := getFsDevice(dir)
+	if err != nil {
+		return err
+	}
+
+	err = xqa.applyQuota(volDevice, quota, *fsGroup)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Check if the empty dir is on an XFS filesystem:
+func (xqa *xfsQuotaApplicator) isXfs(dir string) bool {
+	args := []string{"-f", "-c", "'%T'", dir}
+	outBytes, err := exec.Command("stat", args...).Output()
+	out := string(outBytes)
+	if err != nil {
+		glog.Errorf("error checking filesystem type for %s: %s", dir, err)
+		return false
+	}
+	if strings.Contains(out, "xfs") {
+		return true
+	}
+	return false
+}
+
+func (xqa *xfsQuotaApplicator) applyQuota(volDevice string, quotaInMb int, fsGroupId int64) error {
+	args := []string{"-x", "-c",
+		fmt.Sprintf("limit -g bsoft=%dm bhard=%dm %d", quotaInMb, quotaInMb, fsGroupId),
+		volDevice,
+	}
+
+	// xfs_quota is very happy to fail but return a success code, likely due to it's
+	// interactive shell approach. Grab stderr, if we see anything written to it we'll
+	// consider this an error.
+	cmd := exec.Command("xfs_quota", args...)
+	var stderr bytes.Buffer
+	cmd.Stderr = &stderr
+
+	err := cmd.Run()
+	if err != nil {
+		glog.Errorf("Error running xfs_quota %s: %s", args, err)
+		return err
+	}
+	if stderr.Len() > 0 {
+		glog.Error("Error running xfs_quota: ", string(stderr.String()))
+		return errors.New("xfs_quota wrote to stderr")
+	}
+
+	glog.V(1).Infof("##################")
+	glog.V(1).Infof("XFS quota applied: xfs_quota %s", args)
+	glog.V(1).Infof("##################")
+	return nil
+}
+
+// Returns the fs device for a given emptydir path. To do this we
+// run df on the emptydir path, returning a header line and the line we're
+// interested in. The first string token in that line will be the device name.
+func getFsDevice(dir string) (string, error) {
+	outBytes, err := exec.Command("df", dir).Output()
```

Comment:
If you do `df --output=source $dir` you can make stronger assumptions about the output